lib/puppet/provider/snmp_notification_receiver/nxapi.rb

```diff
@@ -0,0 +1,164 @@
+  end
+
+  def self.properties_get(snmpnotificationreceiver_name, v)
+    debug "Checking instance, SnmpNotificationReceiver #{snmpnotificationreceiver_name}"
+
+    if v.version.eql?('1')
+      version = 'v1'
+    elsif v.version.eql?('2c')
+      version = 'v2'
+    elsif v.version.eql?('3')
+      version = 'v3'
+    else
+      version = nil
+    end
+
+    current_state = {
+      ensure:           :present,
+      name:             snmpnotificationreceiver_name,
+      port:             v.port,
+      username:         v.username,
+      version:          version,
+      type:             v.type,
+      security:         v.security,
+      vrf:              v.vrf,
+      source_interface: v.source_interface,
+    }
+    new(current_state)
+  end # self.properties_get
+
+  def self.instances
+    snmpnotificationreceivers = []
+    Cisco::SnmpNotificationReceiver.receivers.each do |snmpnotificationreceiver_name, v|
+      snmpnotificationreceivers << properties_get(snmpnotificationreceiver_name, v)
+    end
+
+    snmpnotificationreceivers
+  end
+
+  def self.prefetch(resources)
+    snmpnotificationreceivers = instances
+
+    resources.keys.each do |id|
+      provider = snmpnotificationreceivers.find { |snmpnotificationreceiver| snmpnotificationreceiver.name.to_s == id.to_s }
+      resources[id].provider = provider unless provider.nil?
+    end
+  end # self.prefetch
+
+  def exists?
+    @property_hash[:ensure] == :present
+  end
+
+  def create
+    @property_flush[:ensure] = :present
+  end
+
+  def destroy
+    @property_flush[:ensure] = :absent
+  end
+
+  def validate
+    required = []
+    REQUIRED_PROPS.each do |prop|
+      required << prop unless @resource[prop]
+    end
+    fail ArgumentError,
+         "You must specify the following properties: #{required}" \
+           unless required.empty?
+
+    fail ArgumentError,
+         "The 'type' property only supports a setting of 'traps' when 'version' is set to '1'" \
+            if !@resource[:type].eql?('traps') && @resource[:version].eql?(:v1)
+
+    fail ArgumentError,
+         "The 'security' property is only supported when 'version' is set to '3'" \
+            if @resource[:security] && !@resource[:version].eql?(:v3)
+  end
+
+  def munge_flush(val)
+    if val.is_a?(String) && val.eql?('unset')
+      nil
+    elsif val.is_a?(Integer) && val.eql?(-1)
+      nil
+    elsif val.is_a?(Symbol) && val.eql?(:true)
+      true
+    elsif val.is_a?(Symbol) && val.eql?(:false)
+      false
+    elsif val.is_a?(Symbol)
+      val.to_s
+    elsif val.is_a?(Integer)
+      val.to_s
+    else
+      val
+    end
+  end
+
+  def munge_version(_va)
+    if @resource[:version].eql?(:v1)
+      '1'
+    elsif @resource[:version].eql?(:v2)
+      '2c'
+    elsif @resource[:version].eql?(:v3)
+      '3'
+    end
+  end
+
+  def flush # rubocop:disable Metrics/CyclomaticComplexity
+    @snmpnotificationreceiver.destroy if @snmpnotificationreceiver
```

Comment:
It is idempotent.

By the time that this line is executed, the prefect has already been run, and the resources are puppet resources.

The flush method only runs if there is a property update.

Perhaps I have misunderstood your question, or I am missing something here. It would be appreciated if you could provide a manifest example or something similar, thanks.